### PR TITLE
CANS-1006: backend API for Assessment Comparison table

### PR DESCRIPTION
**Pull request by LeonidMarushevskyi**
## JIRA

https://osi-cwds.atlassian.net/browse/CANS-1006

## Description

New API resource added /clients/{id}/assessment_comparison which returns assessments comparison in time.

## Tests
- [ ] I used TDD to develop the feature
- [x] I have included unit tests
- [x] I have included new acceptance tests or modified existing ones
- [ ] I have included other tests

## Types of changes

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I ran all my tests locally which were green.
- [ ] My code adds no new issues to Code Climate
- [ ] I ran all the linters and static analyzers for the project (SonarQube, eslint, rubocop, etc).
- [ ] My code is in a stable state ready to be deployable, but not necessarily complete.

### CODE REVIEW

File: src/main/java/gov/ca/cwds/cans/dao/AssessmentDao.java [hunk 1]
```diff
@@ -128,6 +129,17 @@ private void revertCountyAndCaseIdToInitialValue(Assessment assessment) {
         .list();
   }
 
+  public Collection<Assessment> getLatestAssessmentsByClientId(final String clientIdentifier, final Short limit) {
+    final Session session = grabSession();
+    Require.requireNotNullAndNotEmpty(clientIdentifier);
+    final Short limitOrDefault = limit != null ? limit : 5;
+    return session
+        .createNamedQuery(NQ_LATEST_FOR_CLIENT_WITH_LIMIT, Assessment.class)
```
Comment (DmitriyRudenko) [inline src/main/java/gov/ca/cwds/cans/dao/AssessmentDao.java]: why don't use setMaxResults()?

File: src/main/java/gov/ca/cwds/cans/domain/dto/facade/AssessmentComparisonDto.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  @Data
+  @Accessors(chain = true)
+  @EqualsAndHashCode()
+  @ToString()
+  public static class DomainInTime {
+    private String code;
+    private Boolean underSix;
+    private Boolean aboveSix;
+    private Boolean isCaregiverDomain;
```
Comment (DmitriyRudenko) [inline src/main/java/gov/ca/cwds/cans/domain/dto/facade/AssessmentComparisonDto.java]: isn't just 1 flag (aboveSix or aboveSix) enough?

File: src/main/java/gov/ca/cwds/cans/domain/entity/Assessment.java [hunk 3]
```diff
@@ -77,6 +78,16 @@
...
+@NamedNativeQuery(
+    name = NQ_LATEST_FOR_CLIENT_WITH_LIMIT,
+    query =
+        "SELECT a.* FROM {h-schema}assessment a "
+            + " INNER JOIN {h-schema}person p ON p.id = a.person_id"
+            + " WHERE p.external_id = :" + PARAM_CLIENT_IDENTIFIER
+            + " AND status <> 'DELETED'"
+            + " ORDER by a.event_date desc, a.id"
+            + " LIMIT :" + PARAM_LIMIT,
+    resultClass = Assessment.class)
```
Comment (DmitriyRudenko) [inline src/main/java/gov/ca/cwds/cans/domain/entity/Assessment.java]: why native query?

File: src/test/java/gov/ca/cwds/cans/rest/resource/AssessmentComparisonResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    Response response =
+        clientTestRule
+            .withSecurityToken(AUTHORIZED_USER_NAPA)
+            .target(API.CLIENTS + SLASH + CLIENT_IDENTIFIER + SLASH + ASSESSMENT_COMPARISON)
+            .request(MediaType.APPLICATION_JSON_TYPE)
+            .get();
+    Assert.assertThat(response.getStatus(), equalTo(HttpStatus.SC_OK));
+    final AssessmentComparisonDto responseDto = response.readEntity(AssessmentComparisonDto.class);
+    final String expectedDto = fixture(ASSESSMENT_COMPARISON_5);
+    assertEqualsResponse(expectedDto, transformDTOtoJSON(responseDto), "assessmentId");
```
Comment (DmitriyRudenko) [inline src/test/java/gov/ca/cwds/cans/rest/resource/AssessmentComparisonResourceTest.java]: Why don't use JSONAssert?

File: src/test/java/gov/ca/cwds/cans/test/util/AssertResponseHelper.java [hunk 5]
```diff
@@ -16,15 +17,43 @@ private AssertResponseHelper() {}
...
+      removeExclusion(actualMap, excludedProperty);
+    }
+  }
+
+  private static void removeExclusion(Map<String, Object> map, String excludedProperty) {
+    for (Map.Entry<String, Object> entry : map.entrySet()) {
+      String key = entry.getKey();
+      Object value = entry.getValue();
+      if (value instanceof String || value instanceof Integer || value instanceof Boolean) {
+        if (key.contains(excludedProperty)) {
```
Comment (DmitriyRudenko) [inline src/test/java/gov/ca/cwds/cans/test/util/AssertResponseHelper.java]: value could be double/float, enum or something else, I think this method should remove props if type is not List or Map. Check for primitive types is redundant.

File: src/test/java/gov/ca/cwds/cans/rest/resource/AssessmentComparisonResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  public static final String ASSESSMENT_ABOVE_SIX_COMPLETED =
+      "fixtures/assessment/assessment-above-six-complete.json";
+  public static final String CLIENT_FIXTURE = "fixtures/client-of-0Ki-rw-assignment.json";
+  public static final String ASSESSMENT_COMPARISON_5 = "fixtures/assessment-comparison/5-assessment-comparison.json";
+  public static final String CLIENT_IDENTIFIER = "O9kIYi80Ki";
+
+
+  @Test
+  public void doGetClientAssessmentComparison_success() throws Exception {
+    postAssessmentAndCheckStatus(
```
Comment (denys-davydov) [inline src/test/java/gov/ca/cwds/cans/rest/resource/AssessmentComparisonResourceTest.java]: Can't find if you clean up the posted assessments after the tests run.

File: src/test/resources/fixtures/assessment-comparison/5-assessment-comparison.json [hunk 7]
```diff
@@ -0,0 +1,3505 @@
+{
```
Comment (denys-davydov) [inline src/test/resources/fixtures/assessment-comparison/5-assessment-comparison.json]: These 3 JSON files are huge and are not manageable, is it possible to make them shorter?

File: src/test/java/gov/ca/cwds/cans/rest/resource/AssessmentComparisonResourceTest.java (hunk 6 above)
Comment (LeonidMarushevskyi) [reply src/test/java/gov/ca/cwds/cans/rest/resource/AssessmentComparisonResourceTest.java]: It is in AbstractFunctionalTest -> postAssessmentAndCheckStatus -> checkStatus

File: src/test/resources/fixtures/assessment-comparison/5-assessment-comparison.json (hunk 7 above)
Comment (LeonidMarushevskyi) [reply src/test/resources/fixtures/assessment-comparison/5-assessment-comparison.json]: JSONs contain realistic completed assessments for different age groups and assessment comparison for 5 assessments history. I believe it has value to have realistic test data.

File: src/test/java/gov/ca/cwds/cans/rest/resource/AssessmentComparisonResourceTest.java (hunk 4 above)
Comment (LeonidMarushevskyi) [reply src/test/java/gov/ca/cwds/cans/rest/resource/AssessmentComparisonResourceTest.java]: I don't see JSONAssert supports exclusions. AssessmentId is dynamic so I need to exclude it from the comparison.

File: src/main/java/gov/ca/cwds/cans/domain/entity/Assessment.java (hunk 3 above)
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/domain/entity/Assessment.java]: In style with other similar queries in Assessment.java.

File: src/main/java/gov/ca/cwds/cans/domain/dto/facade/AssessmentComparisonDto.java (hunk 2 above)
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/domain/dto/facade/AssessmentComparisonDto.java]: Domain can have both underSix and aboveSix true at the same time if domain used for both age ranges.

File: src/test/java/gov/ca/cwds/cans/test/util/AssertResponseHelper.java (hunk 5 above)
Comment (LeonidMarushevskyi) [reply src/test/java/gov/ca/cwds/cans/test/util/AssertResponseHelper.java]: Agree. Fixed.

File: src/main/java/gov/ca/cwds/cans/dao/AssessmentDao.java (hunk 1 above)
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/dao/AssessmentDao.java]: Fixed

**Comment by LeonidMarushevskyi** [comment]
Retest this please

**Pull request merged**